L74-java-channel-creds.md

```diff
@@ -0,0 +1,485 @@
+* Last updated: 2020-06-17
+* Discussion at: https://groups.google.com/g/grpc-io/c/GNU9svnDsQs/m/kG6XsUy9BwAJ
+
+## Abstract
+
+Develop an API that separates transport credential construction from
+channel/server construction.
+
+This is to allow:
+
+*   implementing new transport credentials without creating a new
+    channel/server type, which allows users of the API to separate concerns
+*   providing stable APIs for advanced credential configuration like mutual
+    TLS, without polluting the general-purpose ManagedChannelBuilder API with
+    many overloads that don't apply to all transports
+*   composition of credentials, like necessary for xDS where a fallback
+    credential needs to be provided by the user
+*   aligning with C core and Go implementations
+
+_The design focuses on channel credentials, since they are the most complex and
+most discussed._ Server credentials will mostly just mirror the channel
+credentials; they suffer from most of the same problems but at a smaller
+scale.
+
+## Background
+
+Prior to gRPC v0.12.0, C core had a "unified" `Credential` type that was
+simultaneously a call and channel credential. This was the design proposed for
+all languages. However, the Java implementation could not determine how such an
+API could be implemented for Java with its multiple transports. The designer of
+the API had too little bandwidth (it was a _very_ busy time), so Java (and Go)
+developed its own API. Java used interceptors (which no other language had) and
+credential-specific configuration on the ManagedChannelBuilder.
+
+In gRPC v0.12.0, C core swapped to a “split” CallCredential and
+ChannelCredential. ChannelCredentials could contain CallCredentials. At this
+point it was not a priority for Java to investigate the new API; Java was happy
+they didn’t use the previous API, such that they didn’t have an API migration.
+
+gRPC Java 0.15.0 added CallCredentials to replace the interceptor-based call
+credentials. At this point Java had looked into the full “split” credential
+design, but decided it was too hard to implement at that moment. Java worked to
+stabilize the `ManagedChannelBuilder.usePlaintext()` and
+`useTransportSecurity()` APIs.
+
+Go has split call credentials (`credentials.PerRPCCredentials`) and channel
+credentials (`credentials.TransportCredentials`), but with a separate type to
+combine them (`credentials.Bundle`). The split credentials significantly
+predate `Bundle`.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+`ChannelCredentials` will be introduced, which can be provided when
+constructing a `ManagedChannelBuilder':
+
+```java
+// instead of:
+channel = ManagedChannelBuilder.forTarget("example.com")
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
```

Comment:
I've sent out the changes for this proposal in https://github.com/grpc/grpc-java/pull/7294 .

One thing came out of that when using the unstable ChannelCredentials as provided by the transports: they won't have constructors. I don't care _too_ much about users that use the classes in io.grpc.netty or io.grpc.okhttp directly, since they are few and can "deal with it." But it is a bigger concern for more user-facing things like GoogleDefaultChannelCredentials. We have a few of those today and will get more, like XdsChannelCredentials. That could cause "normal users" to see a mix of both construction styles.